### PR TITLE
LC-399 make sure Worker and WorkerIndexer terminate when a startup exception is encountered

### PR DESCRIPTION
**Pull request by rseitz**
When starting Worker or WorkerIndexer at the command line, the java process would not terminate if there was an error in certain parts of the config file, including the pipeline section and/or the kafka section. This was happening because at least one thread had been started before attempting to parse those sections of the config. When the parse error was encountered, there was no logic in place to cleanly stop all threads that might have been started. The lingering threads were preventing the process from terminating. This PR addresses that issue by catching exceptions encountered while starting Worker and WorkerIndexer threads and, in case an exception is encountered, calling stop() on the corresponding thread pool.

Update: after a closer look, it turned out the issue with WorkerIndexer not terminating was being caused by a lingering thread that got created by instantiating a java.util.Timer as an instance variable, and in some error, cases not calling cancel() on the timer. This problem was fixed by delaying timer instantiation to the WorkerIndexer.start() method and passing a flag to make the timer threads "daeomon threads" that won't prevent JVM termination. The earlier code changes in this PR to call stop() in error scenarios are still beneficial.

### CODE REVIEW

File: lucille-core/src/main/java/com/kmwllc/lucille/core/WorkerIndexerPool.java [hunk 1]
```diff
@@ -49,12 +49,26 @@ public void start() throws Exception {
...
+      try {
+        WorkerIndexer workerIndexer = new WorkerIndexer();
+        workerIndexer.start(config, pipelineName, bypassSearchEngine, idSet);
+        workerIndexers.add(workerIndexer);
+      } catch (Exception e) {
+        log.error("Exception caught when starting WorkerIndexer thread {}; aborting", i+1, e);
+        try {
+          stop();
+        } catch (Exception e2) {
+          log.error("Exception caught when attempting to stop WorkerIndexer threads because of a startup problem", e);
```
Comment (akul-sethi) [inline lucille-core/src/main/java/com/kmwllc/lucille/core/WorkerIndexerPool.java:63]: I believe this should be e2

File: lucille-core/src/main/java/com/kmwllc/lucille/core/WorkerPool.java [hunk 2]
```diff
@@ -53,11 +53,25 @@ public void start() throws Exception {
...
-      threads.add(Worker.startThread(config, messenger, pipelineName, metricsPrefix));
+      try {
+        WorkerMessenger messenger = workerMessengerFactory.create();
+        threads.add(Worker.startThread(config, messenger, pipelineName, metricsPrefix));
+      } catch (Exception e) {
+        log.error("Exception caught when starting Worker thread {}; aborting", i+1);
+        try {
+          stop();
+        } catch (Exception e2) {
+          log.error("Exception caught when attempting to stop Worker threads because of a startup problem", e);
```
Comment (akul-sethi) [inline lucille-core/src/main/java/com/kmwllc/lucille/core/WorkerPool.java:66]: i believe it should be e2

**Review by akul-sethi — APPROVED** [review]
looks good to me!

File: lucille-core/src/main/java/com/kmwllc/lucille/core/WorkerIndexerPool.java (hunk 1 above)
Comment (rseitz) [reply lucille-core/src/main/java/com/kmwllc/lucille/core/WorkerIndexerPool.java]: thx -- fixed

File: lucille-core/src/main/java/com/kmwllc/lucille/core/WorkerPool.java (hunk 2 above)
Comment (rseitz) [reply lucille-core/src/main/java/com/kmwllc/lucille/core/WorkerPool.java]: thx -- fixed

**Pull request merged**